scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -142,62 +127,259 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+{
+    node.split_dim = -1;
+    node.children = end_idx - start_idx;
+    node.split = 0.;
+    node.start_idx = start_idx;
+    node.end_idx = end_idx;
+    node.less = node.greater = nullptr;
+    node._less = node._greater = 0;
+}
+
+static ckdtree_intp_t
+build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+      double *maxes, double *mins,
+      const int _median, const int _compact,
+      std::vector<ckdtreenode>& buf)
+{
+    /* put a new node into the node stack */
+    const ckdtree_intp_t node_index = buf.size();
+    ckdtree_intp_t _less, _greater = 0;
+    buf.emplace_back();
+    ckdtreenode *node = &buf.back();
+    init_ckdtreenode(*node, start_idx, end_idx);
+
+    if (end_idx-start_idx <= self->leafsize) {
+        /* below brute force limit, return leafnode */
+        node->split_dim = -1;
+        return node_index;
+    }
+    else {
+
+        const ckdtree_intp_t p = split_and_partition(self, start_idx, end_idx, maxes, mins, _median, _compact, node);
+        if(p < 0)
+            return node_index; // split_and_partition decided to have a leaf node
+
+        const double split = node->split;
+        const ckdtree_intp_t d = node->split_dim;
+
         if (CKDTREE_LIKELY(_compact)) {
-            _less = build(self, start_idx, p, maxes, mins, _median, _compact);
-            _greater = build(self, p, end_idx, maxes, mins, _median, _compact);
+            _less = build(self, start_idx, p, maxes, mins, _median, _compact, buf);
+            _greater = build(self, p, end_idx, maxes, mins, _median, _compact, buf);
         }
         else
         {
-            std::vector<double> tmp(m);
-            double *mids = &tmp[0];
+            std::vector<double> mids(self->m);
 
-            for (i=0; i<m; ++i) mids[i] = maxes[i];
+            std::copy_n(maxes, mids.size(), mids.begin());
             mids[d] = split;
-            _less = build(self, start_idx, p, mids, mins, _median, _compact);
+            _less = build(self, start_idx, p, mids.data(), mins, _median, _compact, buf);
 
-            for (i=0; i<m; ++i) mids[i] = mins[i];
+            std::copy_n(mins, mids.size(), mids.begin());
             mids[d] = split;
-            _greater = build(self, p, end_idx, maxes, mids, _median, _compact);
+            _greater = build(self, p, end_idx, maxes, mids.data(), _median, _compact, buf);
         }
 
-        /* recompute n because std::vector can
-         * reallocate its internal buffer
-         */
-        root = tree_buffer_root(self->tree_buffer);
-        n = root + node_index;
+        node = &buf[node_index];
         /* fill in entries */
-        n->_less = _less;
-        n->_greater = _greater;
-        n->less = root + _less;
-        n->greater = root + _greater;
-        n->split_dim = d;
-        n->split = split;
+        node->_less = _less;
+        node->less = buf.data() + node->_less;
+        node->_greater = _greater;
+        node->greater = buf.data() + node->_greater;
```

Comment:
```node->split_dim``` and ```node->split``` are not filled in

Reply:
They are exposed to Python, so they must be saved, although they are no longer needed for the kd-tree.

Reply:
Both `build` functions call `split_and_partition`, which takes care of this.

Reply:
Ok